{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Factor GRM and REML Analysis\n",
    "\n",
    "This notebook demonstrates how to:\n",
    "1. Preprocess PLINK bfiles and compute Factor GRMs (heterozygous, homozygous, and interaction terms).\n",
    "2. Perform REML analysis using GCTA.\n",
    "3. Parse and save the REML results for further analysis.\n",
    "\n",
    "Follow the steps below to reproduce the analysis."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# Step 0: Preprocess PLINK bfiles\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Before proceeding with GRM computation and REML analysis, ensure that your PLINK bfiles are preprocessed.  \n",
    "This includes quality control (QC), aligning to the major reference allele, and generating allele frequency files.\n",
    "\n",
    "Refer to the `plink.ipynb` notebook for detailed instructions on preprocessing PLINK bfiles.  \n",
    "The preprocessing pipeline in `plink.ipynb` performs the following steps:\n",
    "1. **Quality Control (QC)**: Filters samples and variants based on criteria such as MAF, HWE, and missingness.\n",
    "2. **Alignment**: Aligns the dataset to the major reference allele.\n",
    "3. **Allele Frequency Calculation**: Computes allele frequencies for the dataset.\n",
    "\n",
    "Once preprocessing is complete, the output files will be ready for GRM computation in this notebook."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# Step 1: Set up the environment\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Project root path: /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA\n"
     ]
    }
   ],
   "source": [
    "# Define the root path of the project and import necessary modules\n",
    "import ipynbname\n",
    "import os, sys\n",
    "\n",
    "# Automatically detect the notebook's directory\n",
    "notebook_path = ipynbname.path()\n",
    "notebook_dir = os.path.dirname(notebook_path)\n",
    "ROOT_PATH = os.path.abspath(os.path.join(notebook_dir, '..'))\n",
    "\n",
    "# Add the project root to the Python path\n",
    "sys.path.insert(0, os.path.join(ROOT_PATH))\n",
    "\n",
    "print(f\"Project root path: {ROOT_PATH}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# Step 2: Compute GRMs\n",
    "---\n",
    "\n",
    "This step calculates GRMs (additive, dominance, and factor) from PLINK bfiles."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCTA Python v0.1.0 initialized\n",
      "Calculating GRM with standardization: add\n",
      "> Load PLINK files\n",
      "|   Loaded 3900 samples and 985 SNPs\n",
      "> Load AFREQ file\n",
      "|   Loaded 985 SNPs allele frequencies\n",
      "|   984 SNPs after MAF thresholding\n",
      "|   Using 984 common SNPs\n",
      "> Calculate GRM\n",
      "|   Loading BED file\n",
      "|   Using 3900 samples and 984 SNPs\n",
      "|   Successfully computed GRM!\n",
      "|   Save: /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/grm/test.majref.add.grm.id\n",
      "|   Save: /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/grm/test.majref.add.grm.bin\n",
      "|   Save: /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/grm/test.majref.add.grm.N.bin\n",
      "|   Successfully saved GRM!\n",
      "GRM saved for add standardization.\n",
      "Calculating GRM with standardization: dom\n",
      "> Load PLINK files\n",
      "|   Loaded 3900 samples and 985 SNPs\n",
      "> Load AFREQ file\n",
      "|   Loaded 985 SNPs allele frequencies\n",
      "|   984 SNPs after MAF thresholding\n",
      "|   Using 984 common SNPs\n",
      "> Calculate GRM\n",
      "|   Using 3900 samples and 984 SNPs\n",
      "|   Successfully computed GRM!\n",
      "|   Save: /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/grm/test.majref.dom.grm.id\n",
      "|   Save: /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/grm/test.majref.dom.grm.bin\n",
      "|   Save: /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/grm/test.majref.dom.grm.N.bin\n",
      "|   Successfully saved GRM!\n",
      "GRM saved for dom standardization.\n",
      "Calculating GRM with standardization: factor\n",
      "> Load PLINK files\n",
      "|   Loaded 3900 samples and 985 SNPs\n",
      "> Load AFREQ file\n",
      "|   Loaded 985 SNPs allele frequencies\n",
      "|   984 SNPs after MAF thresholding\n",
      "|   Using 984 common SNPs\n",
      "> Calculate GRM\n",
      "|   Using 3900 samples and 984 SNPs\n",
      "|   Successfully computed GRM!\n",
      "|   Save: /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/grm/test.majref.het.grm.id\n",
      "|   Save: /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/grm/test.majref.het.grm.bin\n",
      "|   Save: /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/grm/test.majref.het.grm.N.bin\n",
      "|   Save: /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/grm/test.majref.hom.grm.id\n",
      "|   Save: /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/grm/test.majref.hom.grm.bin\n",
      "|   Save: /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/grm/test.majref.hom.grm.N.bin\n",
      "|   Save: /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/grm/test.majref.hethom.grm.id\n",
      "|   Save: /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/grm/test.majref.hethom.grm.bin\n",
      "|   Save: /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/grm/test.majref.hethom.grm.N.bin\n",
      "|   Successfully saved GRM!\n",
      "GRM saved for factor standardization.\n"
     ]
    }
   ],
   "source": [
    "from gcta import GCTA\n",
    "gcta = GCTA()\n",
    "\n",
    "for standardization in [\"add\", \"dom\", \"factor\"]:\n",
    "    print(f\"Calculating GRM with standardization: {standardization}\")\n",
    "    \n",
    "    # Compute GRM\n",
    "    gcta.calculate_grm(\n",
    "        bfile=f\"{ROOT_PATH}/test/bfile/test.majref\",\n",
    "        standardization=standardization,\n",
    "        afreq=f\"{ROOT_PATH}/test/bfile/test.majref.afreq\",\n",
    "        maf_threshold=0.01,\n",
    "    )\n",
    "    \n",
    "    # Save GRM\n",
    "    gcta.save_grm(out_prefix=f\"{ROOT_PATH}/test/grm/test.majref\")\n",
    "    print(f\"GRM saved for {standardization} standardization.\")\n",
    "    \n",
    "    # Remove GRM to save memory\n",
    "    if standardization == \"factor\":\n",
    "        for grm_type in [\"het\", \"hom\", \"hethom\"]:\n",
    "            gcta.remove_grm(grm_type=grm_type)\n",
    "    else:\n",
    "        gcta.remove_grm(grm_type=standardization)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# Step 3: Perform REML Analysis\n",
    "---\n",
    "This step performs REML analysis using GCTA. Ensure the GCTA executable path is correct.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running REML analysis for model: AD\n",
      "*******************************************************************\n",
      "* Genome-wide Complex Trait Analysis (GCTA)\n",
      "* version v1.95.0 Mac\n",
      "* (C) 2010-present, Yang Lab, Westlake University\n",
      "* Please report bugs to Jian Yang <jian.yang@westlake.edu.cn>\n",
      "*******************************************************************\n",
      "Analysis started at 16:56:04 KST on Tue Sep 09 2025.\n",
      "Hostname: JerrysMacPro14.local\n",
      "\n",
      "Accepted options:\n",
      "--reml\n",
      "--mgrm /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/grm/AD.mgrm\n",
      "--pheno /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/test.pheno\n",
      "--out /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/reml/test.majref.AD\n",
      "--reml-no-constrain\n",
      "--reml-no-lrt\n",
      "--thread-num 5\n",
      "\n",
      "Note: the program will be running on 5 threads.\n",
      "\n",
      "Reading phenotypes from [/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/test.pheno].\n",
      "Non-missing phenotypes of 3900 individuals are included from [/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/test.pheno].\n",
      "\n",
      "There are 2 GRM file names specified in the file [/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/grm/AD.mgrm].\n",
      "Reading the GRM from the 1th file ...\n",
      "Reading IDs of the GRM from [/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/grm/test.majref.add.grm.id].\n",
      "3900 IDs are read from [/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/grm/test.majref.add.grm.id].\n",
      "Reading the GRM from [/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/grm/test.majref.add.grm.bin].\n",
      "GRM for 3900 individuals are included from [/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/grm/test.majref.add.grm.bin].\n",
      "Reading the GRM from the 2th file ...\n",
      "Reading IDs of the GRM from [/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/grm/test.majref.dom.grm.id].\n",
      "3900 IDs are read from [/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/grm/test.majref.dom.grm.id].\n",
      "Reading the GRM from [/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/grm/test.majref.dom.grm.bin].\n",
      "GRM for 3900 individuals are included from [/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/grm/test.majref.dom.grm.bin].\n",
      "3900 individuals are in common in these files.\n",
      "\n",
      "Performing  REML analysis ... (Note: may take hours depending on sample size).\n",
      "3900 observations, 1 fixed effect(s), and 3 variance component(s)(including residual variance).\n",
      "Calculating prior values of variance components by EM-REML ...\n",
      "Updated prior values: 0.335932 0.329361 0.344492\n",
      "logL: -964.515\n",
      "Running AI-REML algorithm ...\n",
      "Iter.\tlogL\tV(G1)\tV(G2)\tV(e)\t\n",
      "1\t-962.74\t0.34247\t0.31681\t0.34654\t\n",
      "2\t-961.70\t0.34712\t0.30881\t0.34795\t\n",
      "3\t-961.24\t0.35039\t0.30359\t0.34891\t\n",
      "4\t-961.03\t0.35761\t0.29266\t0.35100\t\n",
      "5\t-960.85\t0.35773\t0.29305\t0.35098\t\n",
      "6\t-960.85\t0.35773\t0.29304\t0.35098\t\n",
      "7\t-960.85\t0.35773\t0.29304\t0.35098\t\n",
      "Log-likelihood ratio converged.\n",
      "\n",
      "Summary result of REML analysis:\n",
      "Source\tVariance\tSE\n",
      "V(G1)\t0.357730\t0.027800\n",
      "V(G2)\t0.293042\t0.020647\n",
      "V(e)\t0.350981\t0.009886\n",
      "Vp\t1.001753\t0.034193\n",
      "V(G1)/Vp\t0.357104\t0.019840\n",
      "V(G2)/Vp\t0.292529\t0.017140\n",
      "\n",
      "Sum of V(G)/Vp\t0.649633\t0.014680\n",
      "\n",
      "Sampling variance/covariance of the estimates of variance components:\n",
      "7.728322e-04\t-1.789059e-06\t-3.328322e-05\t\n",
      "-1.789059e-06\t4.262874e-04\t-2.876074e-05\t\n",
      "-3.328322e-05\t-2.876074e-05\t9.773699e-05\t\n",
      "\n",
      "Summary result of REML analysis has been saved in the file [/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/reml/test.majref.AD.hsq].\n",
      "\n",
      "Analysis finished at 16:56:10 KST on Tue Sep 09 2025\n",
      "Overall computational time: 5.79 sec.\n",
      "REML analysis completed for model: AD\n",
      "Running REML analysis for model: Factor\n",
      "*******************************************************************\n",
      "* Genome-wide Complex Trait Analysis (GCTA)\n",
      "* version v1.95.0 Mac\n",
      "* (C) 2010-present, Yang Lab, Westlake University\n",
      "* Please report bugs to Jian Yang <jian.yang@westlake.edu.cn>\n",
      "*******************************************************************\n",
      "Analysis started at 16:56:10 KST on Tue Sep 09 2025.\n",
      "Hostname: JerrysMacPro14.local\n",
      "\n",
      "Accepted options:\n",
      "--reml\n",
      "--mgrm /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/grm/Factor.mgrm\n",
      "--pheno /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/test.pheno\n",
      "--out /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/reml/test.majref.Factor\n",
      "--reml-no-constrain\n",
      "--reml-no-lrt\n",
      "--thread-num 5\n",
      "\n",
      "Note: the program will be running on 5 threads.\n",
      "\n",
      "Reading phenotypes from [/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/test.pheno].\n",
      "Non-missing phenotypes of 3900 individuals are included from [/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/test.pheno].\n",
      "\n",
      "There are 3 GRM file names specified in the file [/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/grm/Factor.mgrm].\n",
      "Reading the GRM from the 1th file ...\n",
      "Reading IDs of the GRM from [/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/grm/test.majref.het.grm.id].\n",
      "3900 IDs are read from [/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/grm/test.majref.het.grm.id].\n",
      "Reading the GRM from [/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/grm/test.majref.het.grm.bin].\n",
      "GRM for 3900 individuals are included from [/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/grm/test.majref.het.grm.bin].\n",
      "Reading the GRM from the 2th file ...\n",
      "Reading IDs of the GRM from [/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/grm/test.majref.hethom.grm.id].\n",
      "3900 IDs are read from [/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/grm/test.majref.hethom.grm.id].\n",
      "Reading the GRM from [/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/grm/test.majref.hethom.grm.bin].\n",
      "GRM for 3900 individuals are included from [/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/grm/test.majref.hethom.grm.bin].\n",
      "Reading the GRM from the 3th file ...\n",
      "Reading IDs of the GRM from [/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/grm/test.majref.hom.grm.id].\n",
      "3900 IDs are read from [/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/grm/test.majref.hom.grm.id].\n",
      "Reading the GRM from [/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/grm/test.majref.hom.grm.bin].\n",
      "GRM for 3900 individuals are included from [/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/grm/test.majref.hom.grm.bin].\n",
      "3900 individuals are in common in these files.\n",
      "\n",
      "Performing  REML analysis ... (Note: may take hours depending on sample size).\n",
      "3900 observations, 1 fixed effect(s), and 4 variance component(s)(including residual variance).\n",
      "Calculating prior values of variance components by EM-REML ...\n",
      "Updated prior values: 0.261559 0.219738 0.299901 0.333656\n",
      "logL: -1165.72\n",
      "Running AI-REML algorithm ...\n",
      "Iter.\tlogL\tV(G1)\tV(G2)\tV(G3)\tV(e)\t\n",
      "1\t-987.82\t0.20856\t0.15197\t0.33261\t0.33850\t\n",
      "2\t-945.58\t0.18146\t0.11178\t0.36931\t0.34200\t\n",
      "3\t-923.07\t0.16579\t0.08487\t0.40404\t0.34449\t\n",
      "4\t-910.94\t0.15613\t0.06599\t0.43409\t0.34626\t\n",
      "5\t-904.50\t0.14994\t0.05258\t0.45854\t0.34753\t\n",
      "6\t-901.16\t0.14588\t0.04306\t0.47750\t0.34844\t\n",
      "7\t-899.48\t0.14317\t0.03635\t0.49169\t0.34909\t\n",
      "8\t-898.64\t0.14134\t0.03167\t0.50202\t0.34956\t\n",
      "9\t-898.24\t0.13737\t0.02141\t0.52537\t0.35061\t\n",
      "10\t-897.87\t0.13730\t0.02140\t0.52619\t0.35070\t\n",
      "11\t-897.87\t0.13730\t0.02140\t0.52619\t0.35070\t\n",
      "12\t-897.87\t0.13730\t0.02140\t0.52619\t0.35070\t\n",
      "Log-likelihood ratio converged.\n",
      "\n",
      "Summary result of REML analysis:\n",
      "Source\tVariance\tSE\n",
      "V(G1)\t0.137295\t0.013968\n",
      "V(G2)\t0.021401\t0.031635\n",
      "V(G3)\t0.526186\t0.035011\n",
      "V(e)\t0.350697\t0.009780\n",
      "Vp\t1.035579\t0.052353\n",
      "V(G1)/Vp\t0.132578\t0.013346\n",
      "V(G2)/Vp\t0.020666\t0.029808\n",
      "V(G3)/Vp\t0.508108\t0.023587\n",
      "\n",
      "Sum of V(G)/Vp\t0.661352\t0.018890\n",
      "\n",
      "Sampling variance/covariance of the estimates of variance components:\n",
      "1.950974e-04\t5.288433e-05\t-2.392293e-05\t-2.701799e-05\t\n",
      "5.288433e-05\t1.000783e-03\t1.381358e-04\t-3.761293e-06\t\n",
      "-2.392293e-05\t1.381358e-04\t1.225799e-03\t-2.458221e-05\t\n",
      "-2.701799e-05\t-3.761293e-06\t-2.458221e-05\t9.565316e-05\t\n",
      "\n",
      "Summary result of REML analysis has been saved in the file [/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/reml/test.majref.Factor.hsq].\n",
      "\n",
      "Analysis finished at 16:56:20 KST on Tue Sep 09 2025\n",
      "Overall computational time: 9.39 sec.\n",
      "REML analysis completed for model: Factor\n"
     ]
    }
   ],
   "source": [
    "for enc_model in [\"AD\", \"Factor\"]:\n",
    "    print(f\"Running REML analysis for model: {enc_model}\")\n",
    "    \n",
    "    GCTA_PATH = f\"{ROOT_PATH}/gcta64/bin/gcta64\"  # Update this path if necessary\n",
    "    MGRM_FILE = f\"{ROOT_PATH}/test/grm/{enc_model}.mgrm\"\n",
    "    PHENOTYPE_FILE = f\"{ROOT_PATH}/test/test.pheno\"\n",
    "    OUTPUT_PREFIX = f\"{ROOT_PATH}/test/reml/test.majref.{enc_model}\"\n",
    "\n",
    "    gcta.reml(\n",
    "        gcta_path=GCTA_PATH,\n",
    "        mgrm_file=MGRM_FILE,\n",
    "        phenotype_file=PHENOTYPE_FILE,\n",
    "        out_prefix=OUTPUT_PREFIX,\n",
    "        no_contrain=True,\n",
    "        no_lrt=True,\n",
    "        n_threads=5\n",
    "    )\n",
    "    print(f\"REML analysis completed for model: {enc_model}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# Step 4: Parse and Save REML Results\n",
    "---\n",
    "This step parses the REML results and saves them for further analysis.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|   Must be (V1: het, V2: hethom, V3: hom)\n",
      "metric        mean        SE\n",
      "h2(het)   0.135376  0.012859\n",
      "h2(hom)   0.518830  0.018762\n",
      "h2(F)     0.654206  0.014317\n",
      "rho       0.079623  0.117798\n",
      "hom/het   3.832521  0.465894\n",
      "Parsed REML Results:\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "FACTOR_HSQ_FILE = f\"{ROOT_PATH}/test/reml/test.majref.FACTOR.hsq\"\n",
    "df_hsqF = gcta.load_factor_hsq(FACTOR_HSQ_FILE, save=True)\n",
    "\n",
    "# Display the parsed results\n",
    "print(\"Parsed REML Results:\")\n",
    "print(df_hsqF)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## Summary\n",
    "---\n",
    "\n",
    "In this notebook, we:\n",
    "1. Computed Factor GRMs from PLINK bfiles.\n",
    "2. Performed REML analysis using GCTA.\n",
    "3. Parsed and visualized the REML results.\n",
    "\n",
    "You can now use the parsed results for further analysis or modeling. For more details, refer to the documentation."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pygcta",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.23"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
