{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f0c2dfb9",
   "metadata": {},
   "source": [
    "# PLINK Preprocessing Pipeline\n",
    "\n",
    "This notebook demonstrates how to preprocess PLINK bfiles for downstream analysis.  \n",
    "The pipeline includes the following steps:\n",
    "\n",
    "1. **Quality Control (QC)**: Filters samples and variants based on criteria such as:\n",
    "   - Minor Allele Frequency (MAF)\n",
    "   - Hardy-Weinberg Equilibrium (HWE)\n",
    "   - Missingness thresholds for individuals and variants\n",
    "2. **Alignment**: Aligns the dataset to the major reference allele.\n",
    "3. **Allele Frequency Calculation**: Computes allele frequencies for the dataset.\n",
    "\n",
    "The output files from this pipeline will be used as input for GRM computation and REML analysis in `basic_usage.ipynb`."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67663f0a",
   "metadata": {},
   "source": [
    "---\n",
    "# Step 1: Set up the environment\n",
    "---\n",
    "Import necessary modules and define the project root path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b0cd2a75",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Project root path: /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA\n"
     ]
    }
   ],
   "source": [
    "import ipynbname\n",
    "import os, sys\n",
    "\n",
    "# Automatically detect the notebook's directory\n",
    "notebook_path = ipynbname.path()\n",
    "notebook_dir = os.path.dirname(notebook_path)\n",
    "ROOT_PATH = os.path.abspath(os.path.join(notebook_dir, '..'))\n",
    "\n",
    "print(f\"Project root path: {ROOT_PATH}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9578fdb0",
   "metadata": {},
   "source": [
    "---\n",
    "# Step 2: Define helper functions for PLINK commands\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "fe84d6df",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "def build_plink_options(params):\n",
    "    \"\"\"Convert parameter dictionary to PLINK command options.\"\"\"\n",
    "    opts = []\n",
    "    for k, v in params.items():\n",
    "        if v is None:\n",
    "            continue\n",
    "        if isinstance(v, bool):\n",
    "            if v:\n",
    "                opts.append(f\"--{k}\")\n",
    "        else:\n",
    "            opts.append(f\"--{k} {v}\")\n",
    "    return \" \\\\\\n    \".join(opts)  # pretty multi-line formatting\n",
    "\n",
    "def do_qc(plink_path, input_prefix, output_prefix, **kwargs):\n",
    "    \"\"\"Perform Quality Control (QC) using PLINK.\"\"\"\n",
    "    defaults = {\n",
    "        \"keep\": None,\n",
    "        \"maf\": 0.01,\n",
    "        \"hwe\": 1e-6,\n",
    "        \"mind\": 0.1,\n",
    "        \"geno\": 0.1,\n",
    "        \"make-bed\": True,\n",
    "        \"rm-dup\": \"exclude-all\"\n",
    "    }\n",
    "    defaults.update(kwargs)\n",
    "\n",
    "    options_str = build_plink_options(defaults)\n",
    "    qc_command = f\"\"\"{plink_path} \\\\\n",
    "    --bfile {input_prefix} \\\\\n",
    "    {options_str} \\\\\n",
    "    --out {output_prefix}\"\"\"\n",
    "\n",
    "    print(\"Running QC command:\\n\", qc_command)\n",
    "    os.system(qc_command)\n",
    "\n",
    "def do_afreq(plink_path, input_prefix, output_prefix):\n",
    "    \"\"\"Calculate allele frequencies using PLINK.\"\"\"\n",
    "    plink_command = f\"\"\"{plink_path} \\\\\n",
    "    --bfile {input_prefix} \\\\\n",
    "    --freq \\\\\n",
    "    --out {output_prefix}\"\"\"\n",
    "    \n",
    "    print(\"Running allele frequency command:\\n\", plink_command)\n",
    "    os.system(plink_command)\n",
    "\n",
    "def do_align(plink_path, input_prefix, output_prefix):\n",
    "    \"\"\"Align dataset to the major reference allele using PLINK.\"\"\"\n",
    "    plink_command = f\"\"\"{plink_path} \\\\\n",
    "    --bfile {input_prefix} \\\\\n",
    "    --maj-ref force \\\\\n",
    "    --make-bed \\\\\n",
    "    --out {output_prefix}\"\"\"\n",
    "    \n",
    "    print(\"Running alignment command:\\n\", plink_command)\n",
    "    os.system(plink_command)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93e11659",
   "metadata": {},
   "source": [
    "---\n",
    "# Step 3: Define the preprocessing pipeline\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bcdd385a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def qc_pipeline(plink_path, input_path, output_path, fn_prefix, **kwargs):\n",
    "    \"\"\"Run the full preprocessing pipeline.\"\"\"\n",
    "    # Step 1: QC\n",
    "    input_file = f\"{input_path}/{fn_prefix}\"\n",
    "    output_file = f\"{output_path}/tmp.{fn_prefix}\"\n",
    "    do_qc(\n",
    "        plink_path=plink_path,\n",
    "        input_prefix=input_file,\n",
    "        output_prefix=output_file,\n",
    "        **kwargs\n",
    "    )\n",
    "\n",
    "    # Step 2: Align to major reference\n",
    "    input_file = output_file\n",
    "    output_file = f\"{output_path}/{fn_prefix}.majref\"\n",
    "    do_align(\n",
    "        plink_path=plink_path,\n",
    "        input_prefix=input_file,\n",
    "        output_prefix=output_file\n",
    "    )\n",
    "\n",
    "    # Step 3: Allele frequency file\n",
    "    do_afreq(\n",
    "        plink_path=plink_path,\n",
    "        input_prefix=output_file,\n",
    "        output_prefix=output_file\n",
    "    )\n",
    "\n",
    "    # Step 4: Clean temporary files\n",
    "    cleanup_cmd = f\"rm {output_path}/tmp.*\"\n",
    "    print(\"Cleaning up temporary files...\")\n",
    "    os.system(cleanup_cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f3613d92",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting the preprocessing pipeline...\n",
      "Running QC command:\n",
      " /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/plink2 \\\n",
      "    --bfile /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/gcta64/test \\\n",
      "    --maf 0.01 \\\n",
      "    --hwe 1e-06 \\\n",
      "    --mind 0.1 \\\n",
      "    --geno 0.1 \\\n",
      "    --make-bed \\\n",
      "    --rm-dup exclude-all \\\n",
      "    --out /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/bfile/tmp.test\n",
      "PLINK v2.0.0-a.7 M1 (1 Sep 2025)                   cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/bfile/tmp.test.log.\n",
      "Options in effect:\n",
      "  --bfile /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/gcta64/test\n",
      "  --geno 0.1\n",
      "  --hwe 1e-06\n",
      "  --maf 0.01\n",
      "  --make-bed\n",
      "  --mind 0.1\n",
      "  --out /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/bfile/tmp.test\n",
      "  --rm-dup exclude-all\n",
      "\n",
      "Start time: Tue Sep  9 17:00:15 2025\n",
      "16384 MiB RAM detected; reserving 8192 MiB for main workspace.\n",
      "Using up to 8 compute threads.\n",
      "3925 samples (2282 females, 1643 males; 3925 founders) loaded from\n",
      "/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/gcta64/test.fam.\n",
      "1000 variants loaded from\n",
      "/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/gcta64/test.bim.\n",
      "Note: No phenotype data present.\n",
      "Note: Skipping --rm-dup since no duplicate IDs are present.\n",
      "Calculating sample missingness rates... done.\n",
      "25 samples removed due to missing genotype data (--mind).\n",
      "IDs written to\n",
      "/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/bfile/tmp.test.mindrem.id\n",
      ".\n",
      "3900 samples (2259 females, 1641 males; 3900 founders) remaining after main\n",
      "filters.\n",
      "Calculating allele frequencies... done.\n",
      "--geno: 15 variants removed due to missing genotype data.\n",
      "--hwe: 0 variants removed due to Hardy-Weinberg exact test (founders only).\n",
      "0 variants removed due to allele frequency threshold(s)\n",
      "(--maf/--max-maf/--mac/--max-mac).\n",
      "985 variants remaining after main filters.\n",
      "Writing\n",
      "/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/bfile/tmp.test.fam\n",
      "... done.\n",
      "Writing\n",
      "/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/bfile/tmp.test.bim\n",
      "... done.\n",
      "Writing\n",
      "/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/bfile/tmp.test.bed\n",
      "... done.\n",
      "End time: Tue Sep  9 17:00:15 2025\n",
      "Running alignment command:\n",
      " /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/plink2 \\\n",
      "    --bfile /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/bfile/tmp.test \\\n",
      "    --maj-ref force \\\n",
      "    --make-bed \\\n",
      "    --out /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/bfile/test.majref\n",
      "PLINK v2.0.0-a.7 M1 (1 Sep 2025)                   cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/bfile/test.majref.log.\n",
      "Options in effect:\n",
      "  --bfile /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/bfile/tmp.test\n",
      "  --maj-ref force\n",
      "  --make-bed\n",
      "  --out /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/bfile/test.majref\n",
      "\n",
      "Start time: Tue Sep  9 17:00:15 2025\n",
      "16384 MiB RAM detected; reserving 8192 MiB for main workspace.\n",
      "Using up to 8 compute threads.\n",
      "3900 samples (2259 females, 1641 males; 3900 founders) loaded from\n",
      "/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/bfile/tmp.test.fam.\n",
      "985 variants loaded from\n",
      "/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/bfile/tmp.test.bim.\n",
      "Note: No phenotype data present.\n",
      "Calculating allele frequencies... done.\n",
      "Writing\n",
      "/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/bfile/test.majref.fam\n",
      "... done.\n",
      "Writing\n",
      "/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/bfile/test.majref.bim\n",
      "... done.\n",
      "Writing\n",
      "/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/bfile/test.majref.bed\n",
      "... done.\n",
      "End time: Tue Sep  9 17:00:15 2025\n",
      "Running allele frequency command:\n",
      " /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/plink2 \\\n",
      "    --bfile /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/bfile/test.majref \\\n",
      "    --freq \\\n",
      "    --out /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/bfile/test.majref\n",
      "PLINK v2.0.0-a.7 M1 (1 Sep 2025)                   cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2025 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/bfile/test.majref.log.\n",
      "Options in effect:\n",
      "  --bfile /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/bfile/test.majref\n",
      "  --freq\n",
      "  --out /Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/bfile/test.majref\n",
      "\n",
      "Start time: Tue Sep  9 17:00:15 2025\n",
      "16384 MiB RAM detected; reserving 8192 MiB for main workspace.\n",
      "Using up to 8 compute threads.\n",
      "3900 samples (2259 females, 1641 males; 3900 founders) loaded from\n",
      "/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/bfile/test.majref.fam.\n",
      "985 variants loaded from\n",
      "/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/bfile/test.majref.bim.\n",
      "Note: No phenotype data present.\n",
      "Calculating allele frequencies... done.\n",
      "--freq: Allele frequencies (founders only) written to233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899%\n",
      "/Users/jerry/Documents/JerryProject/1.Project/Factor/Analysis/src/pyGCTA/test/bfile/test.majref.afreq\n",
      ".\n",
      "End time: Tue Sep  9 17:00:15 2025\n",
      "Cleaning up temporary files...\n",
      "Preprocessing pipeline completed.\n"
     ]
    }
   ],
   "source": [
    "# Step 4: Run the pipeline\n",
    "PLINK_PATH = f\"{ROOT_PATH}/plink2\"\n",
    "INPUT_PATH = f\"{ROOT_PATH}/gcta64\"\n",
    "OUTPUT_PATH = f\"{ROOT_PATH}/test/bfile\"\n",
    "FN_PREFIX = \"test\"\n",
    "\n",
    "print(\"Starting the preprocessing pipeline...\")\n",
    "qc_pipeline(\n",
    "    plink_path=PLINK_PATH,\n",
    "    input_path=INPUT_PATH,\n",
    "    output_path=OUTPUT_PATH,\n",
    "    fn_prefix=FN_PREFIX\n",
    ")\n",
    "print(\"Preprocessing pipeline completed.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b44d66d7",
   "metadata": {},
   "source": [
    "# Summary\n",
    "\n",
    "In this notebook, we:\n",
    "1. Performed Quality Control (QC) on PLINK bfiles.\n",
    "2. Aligned the dataset to the major reference allele.\n",
    "3. Calculated allele frequencies for the dataset.\n",
    "\n",
    "The preprocessed files are now ready for GRM computation and REML analysis in `basic_usage.ipynb`."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pygcta",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.23"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
